networkit/cpp/centrality/GroupClosenessLocalSearch.cpp

```diff
@@ -0,0 +1,1150 @@
+// networkit-format
+
+#include <algorithm>
+#include <atomic>
+#include <cassert>
+#include <cmath>
+#include <omp.h>
+#include <queue>
+#include <random>
+#include <stdexcept>
+#include <unordered_set>
+#include <utility>
+
+#ifdef __AVX2__
+#include <immintrin.h>
+#include <networkit/auxiliary/AlignedAllocator.hpp>
+#endif // __AVX2__
+
+#include <networkit/auxiliary/Parallel.hpp>
+#include <networkit/auxiliary/Random.hpp>
+#include <networkit/centrality/GroupClosenessGrowShrink.hpp>
+#include <networkit/centrality/GroupClosenessLocalSearch.hpp>
+#include <networkit/distance/Diameter.hpp>
+#include <networkit/graph/BFS.hpp>
+#include <networkit/graph/Dijkstra.hpp>
+
+#include <tlx/container/d_ary_addressable_int_heap.hpp>
+#include <tlx/container/d_ary_heap.hpp>
+#include <tlx/unused.hpp>
+
+namespace NetworKit {
+
+namespace {
+
+template <class WeightType>
+class GroupClosenessLocalSearchImpl final
+    : public GroupClosenessLocalSearch::GroupClosenessLocalSearchInterface {
+
+    static constexpr count K = 16; // Number of packed repetitions using AVX
+    static constexpr float maxInt16 = static_cast<float>(std::numeric_limits<uint16_t>::max());
+    static constexpr WeightType invalidDecr = static_cast<WeightType>(-1);
+    static constexpr WeightType infDistance = std::numeric_limits<WeightType>::max();
+
+public:
+    template <class InputIt>
+    GroupClosenessLocalSearchImpl(const Graph &G, InputIt initGroupFirst, InputIt initGroupLast,
+                                  bool runGrowShrink, count maxIterations);
+
+    void run() override;
+
+private:
+    const Graph *G;
+    const bool runGrowShrink;
+    const count maxIterations;
+
+    // Shortest and 2nd shortest distance between the group to all the nodes
+    std::vector<WeightType> distance, distance2;
+    // Shortest and 2nd shortest distances before a swap
+    std::vector<WeightType> oldDistance, oldDistance2;
+    // Distance vectors used by SSSPs to avoid dynamic memory allocations
+    std::vector<std::vector<WeightType>> distanceGlobal, distance2Global;
+    std::vector<WeightType> farnessIncrease;
+
+    std::vector<node> nearest, nearest2;
+    std::vector<node> oldNearest, oldNearest2;
+    std::vector<std::vector<node>> nearestGlobal, nearest2Global;
+
+    std::vector<std::vector<bool>> visitedGlobal;
+
+    std::vector<node> nodesAtDistance, nodesSortedByDistance;
+    std::vector<std::vector<node>> nodesLeftToExploreAtDistanceGlobal;
+    std::vector<std::pair<node, WeightType>> decreasePerThread;
+
+    std::vector<node> stack, candidatesToAddVec;
+    std::vector<float> approxFarnessDecrease;
+    std::vector<count> sumOfMins;
+    std::vector<edgeweight> sumOfDistancesGreaterThan;
```

Comment:
There are still lots of vectors with no description. If the name is descriptive (`nodesSortedByDistance`), that's fine, but for others (`nodesAtDistance`, `nodesLeftToExploreAtDistanceGlobal`, `oldNearest`, `oldNearest2`, etc.) it would make the code much easier to read if there were comments. (The comments that you added to some of the vectors already help a lot.)